File: perf/benchmarks/micro/set_ops/set_ops.cpp
// set_ops — HashSet insert + membership micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <unordered_set>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 500000;
    auto t0 = std::chrono::steady_clock::now();
    std::unordered_set<int64_t> s;
    s.reserve(static_cast<size_t>(n));
    int64_t seed = 1;
    int64_t mod = n * 2;
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        s.insert(seed % mod);
    }
    int64_t found = 0;
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        if (s.count(seed % mod)) ++found;
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(found));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/micro/set_ops/set_ops.py
# set_ops — HashSet insert + membership micro-benchmark.
#
# Insert N pseudo-random integers (LCG) into a hash set, then test membership
# of N more pseudo-random integers from the same stream. CHECKSUM is the
# count of "found" lookups. Argument: N (default 500_000).

import sys
import time


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 500_000
    t0 = time.perf_counter_ns()
    s: set[int] = set()
    seed = 1
    for _ in range(n):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        s.add(seed % (n * 2))
    found = 0
    for _ in range(n):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        if (seed % (n * 2)) in s:
            found += 1
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{found}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
